### PR TITLE
feat: undo/redo for workflow editor

### LINKED ISSUE #3911: Undo / Redo in Workflow Editor
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Pleas do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

When working in the Workflow Editor, it often happens that I accidentally delete a node. It also happens that I make changes to the workflow and at some point realize that I liked a previous version better. In these situations I have to make extensive adjustments to the workflow, which can be annoying and demotivating.

### 2. Describe the feature you'd like to see

- There is an undo button in the ui, which I can click to undo a change in the workflow editor
- There is a redo button in the ui, which I can click to redo a change in the workflow editor
- I can also use keyboard shortcuts to undo / redo things:

### Keyboard Shortcuts
- **Undo**
  - **Windows/Linux**: `Ctrl` + `Z`
  - **macOS**: `Cmd` + `Z`

- **Redo**
  - **Windows/Linux**: `Ctrl` + `Y` or `Ctrl` + `Shift` + `Z`
  - **macOS**: `Cmd` + `Shift` + `Z` or `Cmd` + `Y`

### 3. How will this feature improve your workflow or experience?

An undo/redo feature saves me having to manually adjust the workflow and I don't have to make backups every now and then if I'm not sure whether the current workflow will work that way or not.

### 4. Additional context or comments
relates to #3548 #3551 

I'll implement it as discussed with the dify team, but will need some help testing it.

### 5. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by perzeuss**
# Description

This resolves #3911 by introducing zundo to create a zustand store for creating a history over changs in the workflow editor.
The workflow store as well as the reactflow store remain untouched.

### UI Elements
Buttons are added  to the UI for both Workflow and Chatflow Editor:
image

### Change History
I also added a way to see all changes and navigate through the history.
image

### Translations
I provided translations for all currently supported languages. However, apart from EN and DE, I had all translations generated by AI.

### Keyboard Shortcuts
Common keyboard shortcuts are also supported. 

- **Undo**
  - **Windows/Linux**: `Ctrl` + `Z`
  - **macOS**: `Cmd` + `Z`

- **Redo**
  - **Windows/Linux**: `Ctrl` + `Y` or `Ctrl` + `Shift` + `Z`
  - **macOS**: `Cmd` + `Shift` + `Z` or `Cmd` + `Y`

## Current Limitations
 - InputChange events in Node Panels do not trigger state changes.
  - This is an issue we need to adress since it also affects the sync workflow darft feature
 - Resizing UI elements does not trigger state changes.
  - But I think we can ignore them anyway

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
 - We might want to add documentation about the limitations
 - We might want to document this feature incl. keyboard shortcuts

# How Has This Been Tested?
I recorded a video of manual testing.
https://github.com/langgenius/dify/assets/11357019/4045181b-3b49-4a10-a15e-5e5c3ca518f3

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by perzeuss** [comment]
We could update the screenshots to keep the docs updated, but I don't have access to the workflows used to make them: https://docs.dify.ai/features/workflow?

File: web/app/components/workflow/hooks/use-workflow-history.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      default:
+        // We do not create a history state for every event.
+        // Some events of reactflow may change things the user would not want to undo/redo.
+        // For example: UI state changes like selecting a node.
+        break
+    }
+  }, [])
+
+  return {
+    store,
```
Comment (perzeuss) [inline web/app/components/workflow/hooks/use-workflow-history.ts]:
```suggestion
    store: workflowHistoryStore,
```

File: web/app/components/workflow/hooks/use-nodes-interactions.ts [hunk 2]
```diff
@@ -178,7 +183,8 @@ export const useNodesInteractions = () => {
       setHelpLineVertical()
       handleSyncWorkflowDraft()
     }
-  }, [handleSyncWorkflowDraft, workflowStore, getNodesReadOnly])
+    saveStateToHistory(WorkflowHistoryEvent.NodeDragStop)
```
Comment (perzeuss) [inline web/app/components/workflow/hooks/use-nodes-interactions.ts]: should be 1 line above inside the if statement

**Comment by perzeuss** [comment]
@zxhlyh @crazywoola ready for review

**Comment by zxhlyh** [comment]
Adding first block in an iteration will not be recorded in the history.

**Comment by perzeuss** [comment]
> Adding first block in an iteration will not be recorded in the history.

fixed

**Comment by zxhlyh** [comment]
image
Can we add some disabled style when history state is empty?

**Comment by perzeuss** [comment]
> image
> Can we add some disabled style when history state is empty?
> 

Yes, I will do it later today.

**Comment by zxhlyh** [comment]
image
The last change. Our product manager said to try to include the action of organizing nodes in the undo and redo functionality as well.

**Review by crazywoola — DISMISSED** [review]
@zxhlyh

**Comment by perzeuss** [comment]
> image The last change. Our product manager said to try to include the action of organizing nodes in the undo and redo functionality as well.

I added it, but to keep it simple I reused "Block moved" instead of introducing "Bocks organized".

**Review by zxhlyh — APPROVED** [review]

**Pull request merged**